### ISSUE TITLE
[google_sign_in][web] Incognito window throws cookies not enabled exception 

### ISSUE THREAD

**Issue by mikron123** [flutter/flutter#59145]
When using an incognito window, and trying to sign in with Google, the following exception is being thrown:

Uncaught PlatformException(google_sign_in, idpiframe_initialization_failed, Cookies are not enabled in current environment.)

Until recently it didn't use to do that!
Is there a way to make it work again?

**Comment by iapicca** [comment]
Hi @mikron123 
Can you please provide your `flutter doctor -v`, 
your `flutter run --verbose` 
and your `pubspec.yaml` 
Thank you

**Comment by mikron123** [comment]
Doctor:

`Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel master, 1.18.0-11.1.pre, on Microsoft Windows [Version 10.0.18362.836], locale en-US)

[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[√] Chrome - develop for the web
[√] Android Studio (version 3.6)
[√] VS Code (version 1.45.1)
[√] Connected device (2 available)
`

Verbose:

`[   +9 ms] executing: [c:\flutter/] git -c log.showSignature=false log -n 1 --pretty=format:%H
[  +69 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%H
[   +1 ms] c2b7342ca470b11cfaad4fbfb094f73aa4c85320
[   +1 ms] executing: [c:\flutter/] git tag --contains HEAD
[ +227 ms] Exit code 0 from: git tag --contains HEAD
[   +2 ms] 1.18.0-11.1.pre
           1.18.0-12.0.pre
           1.18.0-13.0.pre
           1.19.0-0.0.pre
           1.19.0-1.0.pre
           1.19.0-2.0.pre
           1.19.0-3.0.pre
           1.19.0-4.0.pre
           1.19.0-5.0.pre
[  +11 ms] executing: [c:\flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[  +43 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/master
[        ] executing: [c:\flutter/] git ls-remote --get-url origin
[  +37 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[  +77 ms] executing: [c:\flutter/] git rev-parse --abbrev-ref HEAD
[  +35 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] master
[  +48 ms] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[   +9 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[   +1 ms] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[  +17 ms] executing: C:\Users\moko\AppData\Local\Android\Sdk\platform-tools\adb.exe devices -l
[  +41 ms] List of devices attached
[   +5 ms] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[   +2 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[  +63 ms] No devices found
[  +10 ms] "flutter run" took 169ms.

#0      throwToolExit (package:flutter_tools/src/base/common.dart:14:3)
#1      RunCommand.validateCommand (package:flutter_tools/src/commands/run.dart:334:7)
<asynchronous suspension>
#2      FlutterCommand.verifyThenRunCommand (package:flutter_tools/src/runner/flutter_command.dart:776:11)
<asynchronous suspension>
#3      FlutterCommand.run.<anonymous closure> (package:flutter_tools/src/runner/flutter_command.dart:693:33)
<asynchronous suspension>
#4      FlutterCommand.run.<anonymous closure> (package:flutter_tools/src/runner/flutter_command.dart)
#5      AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:29)
#6      _rootRun (dart:async/zone.dart:1184:13)
#7      _CustomZone.run (dart:async/zone.dart:1077:19)
#8      _runZoned (dart:async/zone.dart:1619:10)
#9      runZoned (dart:async/zone.dart:1539:10)
#10     AppContext.run (package:flutter_tools/src/base/context.dart:149:18)
#11     FlutterCommand.run (package:flutter_tools/src/runner/flutter_command.dart:683:20)
#12     CommandRunner.runCommand (package:args/command_runner.dart:197:27)
#13     FlutterCommandRunner.runCommand.<anonymous closure> (package:flutter_tools/src/runner/flutter_command_runner.dart:339:21)
#14     _rootRunUnary (dart:async/zone.dart:1192:38)
#15     _CustomZone.runUnary (dart:async/zone.dart:1085:19)
#16     _FutureListener.handleValue (dart:async/future_impl.dart:141:18)
#17     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:686:45)
#18     Future._propagateToListeners (dart:async/future_impl.dart:715:32)
#19     Future._completeWithValue (dart:async/future_impl.dart:526:5)
#20     _AsyncAwaitCompleter.complete (dart:async-patch/async_patch.dart:36:15)
#21     _completeOnAsyncReturn (dart:async-patch/async_patch.dart:303:13)
#22     FlutterVersion.checkFlutterVersionFreshness (package:flutter_tools/src/version.dart)
#23     _rootRunUnary (dart:async/zone.dart:1192:38)
#24     _CustomZone.runUnary (dart:async/zone.dart:1085:19)
#25     _FutureListener.handleValue (dart:async/future_impl.dart:141:18)
#26     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:686:45)
#27     Future._propagateToListeners (dart:async/future_impl.dart:715:32)
#28     Future._completeWithValue (dart:async/future_impl.dart:526:5)
#29     Future._asyncCompleteWithValue.<anonymous closure> (dart:async/future_impl.dart:560:7)
#30     _rootRun (dart:async/zone.dart:1184:13)
#31     _CustomZone.run (dart:async/zone.dart:1077:19)
#32     _CustomZone.runGuarded (dart:async/zone.dart:979:7)
#33     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1019:23)
#34     _microtaskLoop (dart:async/schedule_microtask.dart:43:21)
#35     _startMicrotaskLoop (dart:async/schedule_microtask.dart:52:5)
#36     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:118:13)
#37     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:169:5)
`

it's weird it used to work a few days ago from Incognito with no problems.. I used Chrome.

**Comment by iapicca** [comment]
Hi @mikron123 
I apologise, I meant `flutter run -d chrome -v`
while triggering the exception
thank you

**Comment by mikron123** [comment]
Thing is that Google sign in doesn't work if you activate it on the browser automatically opened by flatter, since it's Chromium.. so I open another Chrome window and go to the localhost... but then the output is only shown at that browser's console (F12) The output was the one I wrote above about the cookies.

**Comment by TahaTesser** [comment]
hi @mikron123 
Tried the issue with official [example](https://pub.dev/packages/google_sign_in#-example-tab-) with `flutter run -d chrome --web-hostname localhost --web-port 7357` and coped the host and opened in incognito and sign in, there is no exception, haven't enabled in the settings until now

<img width="1116" alt="Screenshot 2020-06-11 at 2 28 51 PM" src="https://user-images.githubusercontent.com/48603081/84366015-e5d9a180-abef-11ea-8e23-e31ae1a1a0b3.png">



<details> 
<summary>flutter doctor -v</summary> 

``` 
[✓] Flutter (Channel dev, 1.19.0-5.0.pre, on Mac OS X 10.15.5 19F101, locale
    en-GB)
    • Flutter version 1.19.0-5.0.pre at /Users/tahatesser/Code/flutter_dev
    • Framework revision c264b70ec3 (3 days ago), 2020-06-08 09:27:34 -0700
    • Engine revision c7f9725521
    • Dart version 2.9.0 (build 2.9.0-13.0.dev 87c5e3612b)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.46.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.11.0

[✓] Connected device (4 available)
    • SM M305F   • 32003c30dc19668f • android-arm64  • Android 10 (API 29)
    • macOS      • macOS            • darwin-x64     • Mac OS X 10.15.5 19F101
    • Web Server • web-server       • web-javascript • Flutter Tools
    • Chrome     • chrome           • web-javascript • Google Chrome
      83.0.4103.97

• No issues found!
``` 

</details> 
Thank you

**Comment by ditman** [comment]
_Incognito now gives you the option to block cookies_

![image](https://user-images.githubusercontent.com/7407478/84303411-9a77b280-ab5f-11ea-81aa-37f1228d2c9d.png)

(As seen [here](https://github.com/flutter/flutter/issues/52338#issuecomment-642181625))

Are you blocking cookies inadvertently when opening a new incognito window?

**Comment by mrmax99** [comment]
I was able to reproduce the issue when that "block third-party cookies" setting was enabled, and it seems to go away when it's disabled.

However, I'm having trouble catching the error when it occurs.  It would be nice to catch the error and handle it rather than hanging.  Is anyone else able to catch this error?  I have try/catch all around the google sign-in related code but it doesn't seem to catch it.

**Comment by ditman** [comment]
@mrmax99 Agreed, when I get to working on this again, I'll make the error catchable. I think the web plugin is swallowing a bunch of errors that should be surfaced better!

**Comment by Producer86** [comment]
@ditman Any progress on this? Some users reporting that "block third-party cookies" is enabled for non-incognito mode by default.
Also, this means that, from now on, if someone wants to keep this option enabled but still use google auth, a different method needs to be implemented?

**Comment by SwissCheese5** [comment]
This is basically the root cause of #52338 except it's not just incognito. Privacy badger blocks third party cookies as well.

**Comment by ditman** [comment]
Hello all!

I've posted a PR to make Google Sign In Web slightly more robust when throwing exceptions from the web, here:

https://github.com/flutter/plugins/pull/2943

Note that the only two methods that are supposed to throw exceptions now are:

`init()` may throw `idpiframe_initialization_failed`. This is called from the core plugin on every call that requires an authentication object (see [_ensureInitialized](https://github.com/flutter/plugins/blob/master/packages/google_sign_in/google_sign_in/lib/google_sign_in.dart#L249))
`signIn()` (`popup_closed_by_user` and [others](https://developers.google.com/identity/sign-in/web/reference#error_codes_2))

Both of those methods now convert the JS errors to a proper `PlatformException`, so the errors should be catchable from the app.

**Comment by ditman** [comment]
I've published v0.9.2 of the google_sign_in_web plugin; now the plugin should properly rethrow the exceptions coming from the JS SDK in a more legible fashion, and with links to the JS SDK docs if more info is needed, as suggested by @mrmax99 [here](https://github.com/flutter/flutter/issues/59145#issuecomment-649982038).

I'm going to close this issue, but feel free to reopen if this is still an issue (there's a central tracking ticket: #52338)
